File: src/com/train/algorithm/search/implementInC++/RestoreIPAddresses.cpp
//
// Created by 罗斯 on 4/10/2022.
//
#include <vector>
#include <string>

using namespace std;

class Solution {
public:
    vector<string> restoreIpAddresses(string s) {
        vector<string> res;
        string ip;
        dfs(0, s, ip, res);
        return res;
    }

private:
    void dfs(int d, string s, string ip, vector<string> &res){
        int l = s.length();
        if (d == 4){
            if (l == 0)
                res.push_back(ip);
            return;
        }

        for (int i = 1; i <= min(3, s[0] == '0' ? 1 : l); ++i) {
            string ss = s.substr(0, i);
            if (i == 3 && stoi(ss) > 255) return;
            dfs(d + 1, s.substr(i), ip + (d == 0 ? "" : ".") + ss, res);
        }
    }
};

File: src/com/train/algorithm/search/implementInPython/RestoreIPAddresses.py
class Solution(object):
    def dfs(self, cur, res, index, temp, count):
        if count > 4:
            return
        if count == 4 and index == len(cur):
            res.append(temp)
        for i in range(1, 4):
            if index + i > len(cur):
                break
            s = cur[index:index + i]
            if (s.startswith("0") and len(s) > 1) or (i == 3 and int(s) >= 256):
                continue
            self.dfs(cur, res, index + i, temp + s + ("." if count < 3 else ""), count + 1)
    def restoreIpAddresses(self, s):
        res = []
        self.dfs(s, res, 0, "", 0)
        return res
